src/emailmessage.cpp

```diff
@@ -1629,3 +1712,58 @@ EmailMessage::CryptoProtocol EmailMessage::cryptoProtocol() const
 {
     return cryptoProtocolForKey(m_signingPlugin, m_signingKeys.value(0, QString()));
 }
+
+typedef QPair<QSharedPointer<QMailMessage>, QMailCryptoFwd::DecryptionResult> DecryptionMessage;
+static DecryptionMessage decryptionHelper(QMailMessage *message)
+{
+    const QMailCryptoFwd::DecryptionResult result =
+        QMailCryptographicServiceFactory::decrypt(*message);
+    return DecryptionMessage(QSharedPointer<QMailMessage>(message), result);
+}
+
+void EmailMessage::decrypt()
+{
+    if (m_msg.isEncrypted()) {
+        const QMailMessagePart &encoded = m_msg.partAt(1);
+        if (!encoded.contentAvailable() && m_cryptedDataLocation.isEmpty()) {
+            m_cryptedDataLocation = encoded.location().toString(true);
+            setEncryptionStatus(EmailMessage::CryptedDataDownloading);
```

Comment:
Here I'm left pondering should decrypt() do downloading, which is a different operation. Maybe good, but just observation.

Reply:
Good point, I thought it was simpler in term of UI, I mean less confusing, if you provide only a " decrypt " button, that does the job, even if it requires to actually download the encrypted content first. The alternative is to first tap on the attachment, and then, retap it to decrypt it...

Reply:
Though the app logic can also trigger the download if needed and have just a "decrypt" button regardless of what's done here?

Reply:
Actually, it may even already be like that. With 4.6 on my device and a different ICU in SDK, I cannot test anymore. I think I remember I kept the attachment row available, so you can tap on it to download and then on the decrypt button. If you tap on the decrypt button directly, it will start the download and you will see it.